File: cses/07_mathematics/2417_counting-coprime-pairs/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    if n < 2:
        # Không đủ hai phần tử thì không có cặp nào.
        sys.stdout.write("0\n")
        return

    # Đọc các giá trị, đồng thời tìm maxv để giới hạn kích thước sàng.
    vals = data[1:1 + n]
    maxv = 0
    ivals = [0] * n
    for i in range(n):
        v = int(vals[i])
        ivals[i] = v
        if v > maxv:
            maxv = v

    # freq[v] = số lần giá trị v xuất hiện.
    freq = [0] * (maxv + 1)
    for v in ivals:
        freq[v] += 1

    N = maxv

    # Sàng Eratosthenes: đánh dấu số nguyên tố đến N.
    sieve = bytearray([1]) * (N + 1)
    sieve[0] = 0
    if N >= 1:
        sieve[1] = 0
    r = int(N ** 0.5)
    for i in range(2, r + 1):
        if sieve[i]:
            sieve[i * i::i] = bytearray(len(range(i * i, N + 1, i)))

    # Hàm Mobius mu[]:
    # mu[m] = (-1)^(số ước nguyên tố phân biệt) nếu m squarefree, ngược lại = 0.
    mu = [1] * (N + 1)
    mu[0] = 0
    for p in range(2, N + 1):
        if sieve[p]:
            # Đổi dấu tất cả bội của p (gộp một ước nguyên tố phân biệt).
            mu[p::p] = [-v for v in mu[p::p]]
            pp = p * p
            if pp <= N:
                # Số chia hết cho p^2 -> không squarefree -> mu = 0.
                mu[pp::pp] = [0] * len(range(pp, N + 1, pp))

    # ans = sum_d mu[d] * C(cnt[d], 2), với cnt[d] = số phần tử chia hết cho d.
    # Chỉ xét d có mu[d] != 0; cnt[d] tính bằng slicing sum (chạy ở tốc độ C).
    ans = 0
    fr = freq
    md = mu
    for d in range(1, N + 1):
        m = md[d]
        if m:
            c = sum(fr[d::d])          # cnt[d]
            if c > 1:
                ans += m * (c * (c - 1) // 2)

    sys.stdout.write(str(ans) + "\n")


main()

File: cses/07_mathematics/2417_counting-coprime-pairs/solution.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    // Đọc n giá trị, đồng thời lưu giá trị lớn nhất để giới hạn kích thước sàng.
    int n;
    std::cin >> n;
    std::vector<int> values(n);
    int maximum_value = 0;
    for (int& value : values) {
        std::cin >> value;
        maximum_value = std::max(maximum_value, value);
    }

    // freq[v] = số lần giá trị v xuất hiện.
    std::vector<int> frequency(maximum_value + 1, 0);
    for (const int value : values) {
        ++frequency[value];
    }

    // Sàng tuyến tính (linear sieve) tính hàm Mobius mu[] đến maximum_value.
    // mu[m] = (-1)^(số ước nguyên tố phân biệt) nếu m squarefree, ngược lại = 0.
    std::vector<int> mobius(maximum_value + 1, 0);
    std::vector<int> primes;
    std::vector<char> composite(maximum_value + 1, false);
    if (maximum_value >= 1) {
        mobius[1] = 1;
    }
    for (int value = 2; value <= maximum_value; ++value) {
        if (!composite[value]) {
            // value là số nguyên tố: có đúng một ước nguyên tố nên mu = -1.
            primes.push_back(value);
            mobius[value] = -1;
        }
        for (const int prime : primes) {
            if (prime > maximum_value / value) {
                break;
            }
            const int product = prime * value;
            composite[product] = true;
            if (value % prime == 0) {
                // product chia hết cho prime^2 -> không squarefree -> mu = 0.
                mobius[product] = 0;
                break;
            }
            // Thêm một ước nguyên tố phân biệt -> đổi dấu mu.
            mobius[product] = -mobius[value];
        }
    }

    // ans = sum_{d} mu[d] * C(cnt[d], 2), với cnt[d] = số phần tử chia hết cho d.
    // Chỉ những d có mu[d] != 0 mới đóng góp.
    long long answer = 0;
    for (int divisor = 1; divisor <= maximum_value; ++divisor) {
        if (mobius[divisor] == 0) {
            continue;
        }
        // Đếm số phần tử chia hết cho divisor bằng cách cộng freq của các bội.
        long long divisible_count = 0;
        for (int multiple = divisor; multiple <= maximum_value;
             multiple += divisor) {
            divisible_count += frequency[multiple];
        }
        // C(cnt, 2) = cnt*(cnt-1)/2 là số cặp cùng chia hết cho divisor.
        answer += static_cast<long long>(mobius[divisor]) *
                  divisible_count * (divisible_count - 1) / 2;
    }

    std::cout << answer << '\n';
    return 0;
}
